### ISSUE TITLE
enquos() and enexprs() should have attribute for auto_names

### ISSUE THREAD

**Issue by hadley** [r-lib/rlang#734]
So that we can tell what names were user supplied and what names were autogenerated.

(This is will help when unpacking unnamed tibble columns)

**Comment by lionel-** [comment]
Maybe this should be a different function?

```r
quos <- quos(foo = bar(), baz(quux), !!letters, .named = TRUE)
auto_label(quos)
#> # A tibble: 3 x 3
#>   expr       name      generated
#>   <quos>     <chr>     <lgl>
#> 1 bar()      foo       FALSE
#> 2 baz(quux)  baz(quux) TRUE
#> 3 <chr>      <chr>     TRUE
```

**Comment by hadley** [comment]
Yeah, I think that would be fine too

**Comment by lionel-** [comment]
This would still be helpful in packages like dplyr or tibble I believe, right @romainfrancois @krlmlr?

**Comment by krlmlr** [comment]
Yes, `dplyr:::dplyr_quosures()` currently calls `enquos()` twice for that reason, it would be significantly faster if called only once.

Perhaps unrelated: Are you planning to move `endots()` to C?

``` r
library(rlang)

basic <- function(...) enquos(...)
ignore_empty <- function(...) enquos(..., .ignore_empty = "all")
named_ignore_empty <- function(...) enquos(..., .named = TRUE, .ignore_empty = "all")

bench::mark(
  basic(a = 1),
  ignore_empty(a = 1),
  named_ignore_empty(a = 1)
)
#> # A tibble: 3 × 6
#>   expression                     min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 basic(a = 1)                20.2µs   21.4µs    46119.        0B     46.2
#> 2 ignore_empty(a = 1)         30.1µs   31.8µs    30842.    12.4KB     49.4
#> 3 named_ignore_empty(a = 1)   38.6µs   40.8µs    23991.    21.2KB     48.1
```

<sup>Created on 2022-05-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by krlmlr** [comment]
We now have `exprs_auto_name()`, and `?dots_list` documents that this is equivalent to `enquos(.named = TRUE)` . Do we still need to take action here?

**Comment by lionel-** [comment]
> Are you planning to move endots() to C?

Yes.

> We now have exprs_auto_name(), and ?dots_list documents that this is equivalent to enquos(.named = TRUE) . Do we still need to take action here?

Yes because we need a way of determining which names are user-supplied and which names are auto-labelled.

**Comment by krlmlr** [comment]
Can't we get that info from the names? Non-empty names are user-supplied, empty names should be auto-labeled? What am I missing?

**Comment by lionel-** [comment]
As I understand this issue is specifically about making the pattern easier without having to use `enquos(.named = TRUE)` and `exprs_auto_name()` in two steps. It is a low priority feature and maybe not needed in rlang.

**Comment by krlmlr** [comment]
Thanks. Did you mean `enquos(.named = FALSE)` and `exprs_auto_name()` ?

**Comment by lionel-** [comment]
yes sorry
